numpyro/infer/kernels.py

```diff
@@ -0,0 +1,338 @@
+from abc import ABC, abstractmethod
+from typing import Callable, List, Dict, Tuple
+
+import jax.numpy as jnp
+import jax.scipy.linalg
+import jax.scipy.stats
+import numpy as np
+import numpy.random as npr
+
+import numpyro.distributions as dist
+from numpyro.util import sqrth, posdef, safe_norm
+
+
+class PrecondMatrix(ABC):
+    @abstractmethod
+    def compute(self, particles: jnp.ndarray, loss_fn: Callable[[jnp.ndarray], float]):
+        """
+        Computes a preconditioning matrix for a given set of particles and a loss function
+        :param particles: The Stein particles to compute the preconditioning matrix from
```

Comment:
Went through and checked papers, all assume kernels `(d,) (d,) -> ...` so each particle will be of shape `(d,)`. I fixed the `ndim >= 2` cases to `ndim == 2` to avoid this confusion.